File: app/routes/utils/shopifyProxyUtils.ts
import crypto from 'crypto';

export function verifyAppProxySignature(query: URLSearchParams, apiSecret: string): boolean {
  const { signature, ...params } = Object.fromEntries(query.entries());

  if (!signature) return false;

  // Sort parameters alphabetically
  const sortedParams = Object.keys(params)
    .sort()
    .reduce((acc, key) => {
      acc[key] = params[key];
      return acc;
    }, {} as Record<string, string>);

  const signatureMessage = Object.keys(sortedParams)
    .map((key) => `${key}=${sortedParams[key]}`)
    .join('');

  const hmac = crypto
    .createHmac('sha256', apiSecret)
    .update(signatureMessage)
    .digest('hex');

  return hmac === signature;
}

File: api/utils/shopify_proxy_utils.py
import hashlib
import hmac
from urllib.parse import parse_qs

def verify_app_proxy_signature(query_params: dict, api_secret: str) -> bool:
    signature = query_params.get("signature")
    if not signature:
        return False

    params = {k: v for k, v in query_params.items() if k != "signature"}
 
    sorted_params = dict(sorted(params.items()))

    signature_message = "&".join(f"{k}={v}" for k, v in sorted_params.items())

    calculated_hmac = hmac.new(
        api_secret.encode('utf-8'),
        signature_message.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()
    
    return hmac.compare_digest(calculated_hmac, signature)